xdsl/parser.py

```diff
@@ -1,1335 +1,1942 @@
+        succ = self.must_parse_optional_successor_list()
+        attrs = self.must_parse_optional_attr_dict()
+        regions = self.must_parse_region_list()
+
+        return args, succ, attrs, regions, None
+
+    def must_parse_optional_successor_list(self) -> list[Span]:
+        if not self.tokenizer.starts_with("("):
+            return []
+        self.must_parse_characters(
+            "(", "Successor list is enclosed in round brackets")
+        successors = self.must_parse_list_of(self.try_parse_block_id,
+                                             "Expected a block-id",
+                                             allow_empty=False)
+        self.must_parse_characters(
+            ")", "Successor list is enclosed in round brackets")
+        return successors
+
+    def must_parse_dialect_type_or_attribute_inner(self, kind: str):
+        if self.tokenizer.starts_with('"'):
+            name = self.try_parse_string_literal()
+            if name is None:
+                self.raise_error(
+                    "Expected string literal for an attribute in generic format here!"
+                )
+            return self.must_parse_generic_attribute_args(name)
+        return super().must_parse_dialect_type_or_attribute_inner(kind)
+
+    def must_parse_generic_attribute_args(self, name: StringLiteral):
+        attr = self.ctx.get_optional_attr(name.string_contents)
+        if attr is None:
+            self.raise_error("Unknown attribute name!", name)
+        if not issubclass(attr, ParametrizedAttribute):
+            self.raise_error("Expected ParametrizedAttribute name here!", name)
+        self.must_parse_characters(
+            '<', 'Expected generic attribute arguments here!')
+        args = self.must_parse_list_of(self.try_parse_attribute,
+                                       'Unexpected end of attribute list!')
+        self.must_parse_characters(
+            '>', 'Malformed attribute arguments, reached end of args list!')
+        return attr(args)
+
+    def must_parse_op_args_list(self) -> list[Span]:
+        self.must_parse_characters(
+            "(", "Operation args list must be enclosed by brackets!")
+        args = self.must_parse_list_of(self.try_parse_value_id_and_type,
+                                       "Expected another bare-id here")
+        self.must_parse_characters(
+            ")", "Operation args list must be closed by a closing bracket")
+        # TODO: check if type is correct here!
+        return [name for name, _ in args]
+
+    def try_parse_type(self) -> Attribute | None:
+        return self.try_parse_attribute()
+
+
+# COMPAT layer so parser_ng is a drop-in replacement for parser:
+
+
+class Source(Enum):
+    XDSL = 1
+    MLIR = 2
+
+
+def Parser(ctx: MLContext,
+           prog: str,
+           source: Source = Source.XDSL,
+           filename: str = '<unknown>',
+           allow_unregistered_ops=False) -> BaseParser:
+    selected_parser = {
```

Comment:
Not sure whether it makes sense to build a dictionary here. Maybe a switch statement?

Reply:
Our tests are running on 3.10 and  3.11, but I am fine with this, if this is widely applied in the community.

Reply:
Yeah I agree with using a dictionary here.
We do have Python 3.10 though, but I would prefer not using the match clause, since `yapf` still doesn't support it, and we don't have a good replacement for `yapf` :/